components/script/dom/path2d.rs

```diff
@@ -59,24 +63,344 @@ impl Path2D {
+                        let cp1 = Point2D::new(*cp1x, *cp1y);
+                        let cp2 = Point2D::new(*cp2x, *cp2y);
+                        Some(compute_arc_to_points(cp0, cp1, cp2, *radius))
+                    },
+                }
+            },
+            PathSegment::MoveTo { x, y } |
+            PathSegment::LineTo { x, y } |
+            PathSegment::Quadratic { x, y, .. } |
+            PathSegment::Bezier { x, y, .. } |
+            PathSegment::Ellipse { x, y, .. } |
+            PathSegment::SvgArc { x, y, .. } => Some(Point2D::new(*x, *y)),
+            PathSegment::ArcTo {
+                cp1x,
+                cp1y,
+                cp2x,
+                cp2y,
+                radius,
+            } => {
+                let last_segment = path.len() - 1;
+                let cp0 = match Path2D::compute_final_arc_to_point(&path[..last_segment]) {
+                    Some(point) => point,
+                    None => Point2D::new(*cp1x, *cp1y),
+                };
+                let cp1 = Point2D::new(*cp1x, *cp1y);
+                let cp2 = Point2D::new(*cp2x, *cp2y);
+                Some(compute_arc_to_points(cp0, cp1, cp2, *radius))
+            },
+        }
+    }
+
+    fn last_point(path: &[PathSegment]) -> Option<Point2D<f32>> {
+        let last_point = path.last()?;
+
+        match last_point {
+            PathSegment::MoveTo { x, y } |
+            PathSegment::LineTo { x, y } |
+            PathSegment::Quadratic { x, y, .. } |
+            PathSegment::Bezier { x, y, .. } |
+            PathSegment::Ellipse { x, y, .. } |
+            PathSegment::SvgArc { x, y, .. } => Some(Point2D::new(*x, *y)),
+            PathSegment::ArcTo { cp1x, cp1y, .. } => Some(
+                Path2D::compute_final_arc_to_point(&path[..path.len() - 1])
+                    .unwrap_or(Point2D::new(*cp1x, *cp1y)),
+            ),
+            PathSegment::ClosePath => {
+                // find first point for the last subpath
+                let first_point_index = {
+                    let mut index = None;
+                    for i in (0..path.len()).rev().skip(1) {
+                        if matches!(path[i], PathSegment::ClosePath) {
+                            index = Some(i);
+                            break;
+                        }
+                    }
+
+                    if let Some(index) = index {
+                        index + 1
+                    } else {
+                        0
+                    }
+                };
+
+                let first_point = path.get(first_point_index)?;
+                match first_point {
+                    PathSegment::MoveTo { x, y } |
+                    PathSegment::LineTo { x, y } |
+                    PathSegment::Quadratic { x, y, .. } |
+                    PathSegment::Bezier { x, y, .. } |
+                    PathSegment::Ellipse { x, y, .. } |
+                    PathSegment::SvgArc { x, y, .. } => Some(Point2D::new(*x, *y)),
```

Comment:
Although to continue my point from https://github.com/servo/servo/pull/37838/files#r2195651271, shouldn't we use first (control) points here?`

Reply:
No at least according to the spec, all of them state to connect the previous point to given point. For example [quadratic curve to](https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-quadraticcurveto) says that as well as `bezierCurve`.

Oh actually, `SvgArc` and `Ellipse` might need their own computation. It seems the `x, y` pair isn't necessarily the end point of the computed arc.

Reply:
After looking back at this, I think we should ignore `SvgArc` because it doesn't look like it's a method used for `Path2D`. I left a question about `Ellipse` on a line above this. With the exception of these segments all the segments should use the `x, y` pair for the new subpath after `closePath` is called.

Reply:
ahhh ok, good to know, thanks